### ISSUE TITLE
Python 3: work with different dictionary walking methods

### ISSUE THREAD

**Issue by josephsl** [nvaccess/nvda#9067]
Hi,

Continuing the journey toward Python 3:

### Is your feature request related to a problem? Please describe.
When walking through a dictionary, Python 2 provides two iteration methods, whereas Python 3 only provides one. Specifically, Python 2 provides a method to return items/keys/values as a list or as an iterator, whereas Python 3 exposes the latter only.

Different parts of NVDA rely on behavioral differences between these two methods. For example, when looking up keys for input management (inputCore), iteritems is used, whereas when looking up voice names in synthesizers (part of synth settings ring), a list of values is used.

### Describe the solution you'd like
There are two solutions:

1. Native/wrappers: for any code fragment that requires a list of dictionary items/keys/values, wrap it around a list call. This means dict.items/keys/values will be strictly for iteration.
2. Six: use methods provided by this shim to return list/iterator of items/keys/values as appropriate.

### Describe alternatives you've considered
None

### Additional context
Not only this affects NVDA Core, parts of add-ons that iterate through a dictionary must be rewritten to handle behavior differences between Python 2 and 3. Add-ons community will be encouraged to adopt the solution from Core once a pull request for this item is merged into master branch.

Thanks.

**Comment by josephsl** [comment]
Hi,

Additional notes:

1. At this point, I recommend adopting the Six module method with a note that it should be ported to native solution after evaluating python 3 transition work.
2. The pull request regarding this feature won't apply to config.conf (Config Manager) because dict.iteritems and dict.items are used for different purposes unless @LeonardDer says it can be safely merged somehow.

Thanks.

**Comment by josephsl** [comment]
Hi,

Actually, upon further reading, it appears Six module does not have an equivalent of dict.items/keys/values for Python 2 (returning lists). So it appears this must be one of those issues that should be solved as part of a master pull request that switches everything to Python 3 at once; this changes if we can find a workaround that preserves compatibility for a while.

Thanks.

**Comment by leonardder** [comment]
> Actually, upon further reading, it appears Six module does not have an equivalent of dict.items/keys/values for Python 2 (returning lists). So it appears this must be one of those issues that should be solved as part of a master pull request that switches everything to Python 3 at once; this changes if we can find a workaround that preserves compatibility for a while.

There is a pretty straight forward workaround.

Instead of `dict.iteritems()`, use `six.iteritems(dict)`
Instead of `dict.items()`, use `list(dict.items())`
The latter will result into a list in python 3. In python 2, it will just pass a list to the constructor of list, which copies it. Not very efficient, but it will certainly suffice.
Alternatively, we can just use `list(six.iteritems(dict))`

>     2. The pull request regarding this feature won't apply to config.conf (Config Manager) because dict.iteritems and dict.items are used for different purposes unless @leonardder says it can be safely merged somehow.

I'm not sure what you mean here. dict.items() isn't used within the config manager. iteritems is defined within AggregatedSection, however since AggregatedSection isn't a dictionary by itself, you can safely rename that iteritems methods to items, I think. Something that ought to be tested, though.

**Comment by josephsl** [comment]
Hi, regarding Config Manager, yep, thanks for correcting me on where to look for these methods. As for coercing an iterator into a list, perhaps one way is to just call list() constructor for now when encountering dict.items/keys/values, with Six module being used in other places. I hope we can find a more consistent solution for readability purposes. Thanks.

**Comment by josephsl** [comment]
Hi,

In the end, native methods will be used with config manager untouched.

Thanks.

**Comment by josephsl** [comment]
Hi,

For future reference: whenever one changes iter* to ones without, make sure to check the type of the class where the method is defined in (hence skipping ConfigManager class). Also, staring from today's commits, explanatory comments will be added (see #7105 comment from Reef for details).

Thanks.

**Comment by leonardder** [comment]
Fixed in #9671
